{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re\n",
    "import pandas as pd\n",
    "import ast\n",
    "import sys\n",
    "import json\n",
    "#import jinja2\n",
    "import argparse\n",
    "import os\n",
    "import numpy as np\n",
    "\n",
    "#%reset_selective karyotype\n",
    "\n",
    "def print_progress_bar(i, max):\n",
    "    '''\n",
    "    i is your current iteration max should be equal to your maximum number of iterations - 1 (otherwise the bar stop at 99%)\n",
    "    50 and 100 are arbutrary. 50 defines the size of the bar in the terminal. 100 makes this a percentage.\n",
    "    The block using this function shoudl be followed by a print(\"\\n\") to skip to the next line.\n",
    "    '''\n",
    "    j = int(i*50/max)\n",
    "    sys.stdout.write(\"\\r\")\n",
    "    sys.stdout.write(\"[%-50s]%d%%\"%('='*j, 100/max*i))\n",
    "    sys.stdout.flush()\n",
    "\n",
    "#https://www.geeksforgeeks.org/python-convert-key-value-string-to-dictionary/\n",
    "\n",
    "def htmlTodataframe(file):\n",
    "    '''\n",
    "    This function transform karyotype.js informations into dataframe, \n",
    "    file = karyotype.js\n",
    "    columns = columns name of karyotype file\n",
    "    '''\n",
    "    \n",
    "    colname = []\n",
    "    if os.path.isfile(file):\n",
    "        print(\"[#INFO] Config file: \"+ os.path.dirname(os.path.realpath(file))+file)\n",
    "        data = []\n",
    "        datas = []\n",
    "        with open(file) as f:\n",
    "            for line in f:\n",
    "                if 'chr' in line:\n",
    "                    proper = (line[:-2].strip())\n",
    "                    r = re.sub('{', '', proper)\n",
    "                    t = re.sub('}', '', r)\n",
    "                    res = dict(map(str.strip, sub.split(':', 1)) for sub in t.split(', ') if ':' in sub)\n",
    "                    #print(res)\n",
    "                    data.append(res)\n",
    "        for values in data:\n",
    "            for k, v in values.items():\n",
    "                if '\"' in v:\n",
    "                    values[k] = v.replace('\"', \"\")\n",
    "            \n",
    "            datas.append(values)\n",
    "        #print(datas)\n",
    "  \n",
    "        for keys in datas[0].keys():\n",
    "            colname.append(keys)\n",
    "        print(\"[#INFO] Config values: \",colname)\n",
    "        df = pd.DataFrame(columns=colname)\n",
    "        \n",
    "        for item in datas:\n",
    "            #print(item)\n",
    "            row = pd.Series(item)\n",
    "            df = df.append(row, ignore_index=True)\n",
    "        return df \n",
    "    else:\n",
    "        print(\"ERROR \"+file+\" does not exist !!!\")\n",
    "        exit()\n",
    "   \n",
    "\n",
    "\n",
    "def dataframeTohtml(outputjs, name, dictargs, boolD, df):\n",
    "    '''\n",
    "    Transform filtered dataframe into htmlfile\n",
    "    output = name of the outputfile  extension js\n",
    "    name = name of the instance in javascript\n",
    "    dictargs = parameters of the instance, for example size of Radius, thickness etc\n",
    "    '''\n",
    "    #print(\"[#INFO] Value of dataframe =\",dataframe)\n",
    "\n",
    "    \n",
    "    tohtml = []\n",
    "    tohtml.append(name)\n",
    "    tohtml.append(dictargs.copy())\n",
    "    \n",
    "    if name != 'ARC_hg19':\n",
    "        name = 'var '+name\n",
    "    if boolD == True:\n",
    "        data = df.to_dict('records')\n",
    "        tohtml.append(data.copy())\n",
    "        jsonString = json.dumps(tohtml)\n",
    "        jsonFile = open(outputjs, 'w+')\n",
    "        jsonFile.write(name+\" = \"+jsonString)\n",
    "        jsonFile.write(';')\n",
    "        jsonFile.close()\n",
    "    \n",
    "    else:\n",
    "        jsonString = json.dumps(tohtml)\n",
    "        jsonFile = open(outputjs, 'w+')\n",
    "        jsonFile.write(name+\" = \"+jsonString)\n",
    "        jsonFile.write(';')\n",
    "        jsonFile.close()\n",
    "            \n",
    "\n",
    "def createMain(dictargs, wanted):\n",
    "    '''\n",
    "    Function to create the main js\n",
    "    '''\n",
    "    mainfile = 'output.html'\n",
    "    \n",
    "    m1 = \"\"\"\n",
    "    <!DOCTYPE html>\n",
    "    <html>\n",
    "        <head>\n",
    "            <meta charset=\"utf-8\">\n",
    "            <title>BioCircos.js</title>\n",
    "        </head>\n",
    "        <body>\n",
    "            <!-- BioCircos.js, Jquery.js and D3.js import -->\n",
    "    \"\"\"\n",
    "    jsPath_lib = '../lib/'\n",
    "    jsFile = ['jquery.js', 'd3.js', 'biocircos-1.1.0.js']\n",
    "    \n",
    "    jsScript = []\n",
    "    for elem in jsFile:\n",
    "        jsScript.append('\\t\\t<script src=\"'+jsPath_lib+elem+'\"></script>')\n",
    "    \n",
    "    m2 = ''' \n",
    "            <!-- Prepare a <div> tag with \"biocircos\" id to set the picture position your will draw in html -->\n",
    "            <div id=\"biocircos\"></div>\n",
    "    <!-- Data configuration -->\n",
    "    '''\n",
    "    jsPath_data = '../config_test/'\n",
    "    jsFile_create = []\n",
    "    jsFile_own = ['test.js', 'SNP04_test.js']\n",
    "    for elem in jsFile_own:\n",
    "        jsFile_create.append('<script src=\"'+jsPath_data+elem+'\"></script>')\n",
    "    \n",
    "    m3 = '''\n",
    "    </script>\n",
    "    <!-- Genome configuration -->\n",
    "    <script>\n",
    "    '''\n",
    "    \n",
    "    genome = [\n",
    "             [\"1\" , 249250621],\n",
    "             [\"2\" , 243199373],\n",
    "             [\"3\" , 198022430],\n",
    "             [\"4\" , 191154276],\n",
    "             [\"5\" , 180915260],\n",
    "             [\"6\" , 171115067],\n",
    "             [\"7\" , 159138663],\n",
    "             [\"8\" , 146364022],\n",
    "             [\"9\" , 141213431],\n",
    "             [\"10\" , 135534747],\n",
    "             [\"11\" , 135006516],\n",
    "             [\"12\" , 133851895],\n",
    "             [\"13\" , 115169878],\n",
    "             [\"14\" , 107349540],\n",
    "             [\"15\" , 102531392],\n",
    "             [\"16\" , 90354753],\n",
    "             [\"17\" , 81195210],\n",
    "             [\"18\" , 78077248],\n",
    "             [\"19\" , 59128983],\n",
    "             [\"20\" , 63025520],\n",
    "             [\"21\" , 48129895],\n",
    "             [\"22\" , 51304566],\n",
    "             [\"X\" , 155270560],\n",
    "             [\"Y\" , 59373566]\n",
    "          ]\n",
    "    \n",
    "    tokeep = []\n",
    "    gen = np.array(genome)\n",
    "    for chroms in gen:\n",
    "        if str(chroms[0]) in wanted:\n",
    "            tokeep.append(list(chroms))\n",
    "    for value in tokeep:\n",
    "        value[1] = int(value[1])\n",
    "    \n",
    "    print(type(tokeep))\n",
    "    varBiocircos = 'var BioCircosGenome = '+str(tokeep) \n",
    "    #print(tokeep)\n",
    "    \n",
    "    BioCircos = '''\n",
    "      BioCircos01 = new BioCircos(ARC_hg19,BACKGROUND01,SNP04_test,BioCircosGenome,{\n",
    "      //Main configuration\n",
    "         target : \"biocircos\",\n",
    "         svgWidth : 900,\n",
    "         svgHeight : 600,\n",
    "         chrPad : 0.04,\n",
    "         innerRadius: 246,\n",
    "         outerRadius: 270,\n",
    "         SNPMouseOverDisplay : true,\n",
    "         SNPMouseOverColor : \"red\",\n",
    "         SNPMouseOverCircleSize : 5,\n",
    "         SNPMouseOverCircleOpacity : 1.0,\n",
    "         SNPMouseOverCircleStrokeColor : \"#F26223\",\n",
    "         SNPMouseOverCircleStrokeWidth : 3,\n",
    "         SNPMouseOverTooltipsHtml01 : \"chr : \",\n",
    "         SNPMouseOverTooltipsHtml02 : \"<br>position : \",\n",
    "         SNPMouseOverTooltipsHtml03 : \"<br>-log10(Pvalue) : \",\n",
    "         SNPMouseOverTooltipsHtml04 : \"<br>rsid : \",\n",
    "         SNPMouseOverTooltipsHtml05 : \"\",\n",
    "         SNPMouseOverTooltipsBorderWidth : 1,\n",
    "         SNPMouseOutDisplay : true,\n",
    "         SNPMouseOutAnimationTime : 700,\n",
    "         SNPMouseOutColor : \"none\",\n",
    "         SNPMouseOutCircleSize : \"none\",\n",
    "         SNPMouseOutCircleOpacity : 1.0,\n",
    "         SNPMouseOutCircleStrokeWidth : 0,\n",
    "      });\n",
    "      BioCircos01.draw_genome(BioCircos01.genomeLength);\n",
    "    '''\n",
    "    \n",
    "    end = '''\n",
    "    </script>\n",
    "        <body style=\"background-color:rgb(176, 196, 222);\">\n",
    "        </body>\n",
    "    </html>\n",
    "    '''\n",
    "    \n",
    "    f = open(mainfile, 'w+')\n",
    "    f.write(m1)\n",
    "    \n",
    "    #Path of configuration files provide by Biocircos\n",
    "    for items in jsScript:\n",
    "        f.write(items+'\\n')\n",
    "    \n",
    "    f.write(m2)\n",
    "        \n",
    "    #path of configuration file created by dataFrametohtml function    \n",
    "    for elems in jsFile_create:\n",
    "        f.write(elems+'\\n')\n",
    "    \n",
    "    for keys in dictargs:\n",
    "        keys = '\"'+keys+'\"'\n",
    "        \n",
    "    f.write(\"<script>\")\n",
    "    f.write(\"\\nvar BACKGROUND01 = ['BACKGROUND01', \"+str(dictargs)+'];')\n",
    "    f.write(m3)\n",
    "    f.write(varBiocircos)\n",
    "    f.write(BioCircos)\n",
    "    f.write(end)\n",
    "    f.close()\n",
    "\n",
    "def selectData(df, chr):\n",
    "    print(\"chr = \", chr)\n",
    "    #print(df.head())\n",
    "    #df_filter = df.loc[df['chr'] == chr]\n",
    "    df_filter = df.loc[df['chr'].isin(chr)]\n",
    "    \n",
    "    return df_filter\n",
    "    \n",
    "        \n",
    "def main():\n",
    "        \n",
    "    karyotype = 'karyotype.human.hg19.js'\n",
    "    \n",
    "    #file ='test.js'\n",
    "    #columns = [\"chr\", \"start\", \"end\", \"color\"]\n",
    "    \n",
    "    SNP = 'SNP04_gwascatalog.js'\n",
    "    #SNP_columns = [\"chr\", \"pos\", \"value\", \"des\", \"color\"]\n",
    "    \n",
    "    configjs = 'SNP04_test.js'\n",
    "    name_config = 'SNP04_test'\n",
    "    \n",
    "    outputjs = 'test.js'\n",
    "    name = 'ARC_hg19'\n",
    "    dictargs = {\n",
    "          \"BginnerRadius\": 205,\n",
    "          \"BgouterRadius\": 153,\n",
    "          \"BgFillColor\": \"#F2F2F2\",\n",
    "          \"BgborderColor\" : \"#000\",\n",
    "          \"BgborderSize\" : 0.3\n",
    "        }\n",
    "    dictargs_config = {\n",
    "      \"maxRadius\": 205,\n",
    "      \"minRadius\": 153,\n",
    "      \"SNPFillColor\": \"#9400D3\",\n",
    "      \"PointType\": \"circle\",\n",
    "      \"circleSize\": 2,\n",
    "      \"rectWidth\": 2,\n",
    "      \"rectHeight\": 2,\n",
    "      \"displaySNPAxis\": \"false\",\n",
    "      \"SNPAxisColor\": \"#B8B8B8\",\n",
    "      \"SNPAxisWidth\": 0.5\n",
    "    }\n",
    "\n",
    "    chr = ['13', '17']\n",
    "    config = selectData(htmlTodataframe(karyotype), chr)\n",
    "    print(config)\n",
    "    dataframeTohtml(outputjs, name, dictargs, boolD=True, df=config)\n",
    "    createMain(dictargs, chr)\n",
    "    \n",
    "def myoptions():\n",
    "    parser = argparse.ArgumentParser()\n",
    "    parser.add_argument(\"-c\", \"--chr\", type = list, default = \"[]\", help = \"list of wanted chr to disp\", dest = 'chr')\n",
    "    \n",
    "if __name__ == '__main__':\n",
    "    args = myoptions()\n",
    "    #main()\n",
    "\n",
    "    #SNP file must have the same length than the genome file \n",
    "#dataframeTohtml(output=output, name=name, dictargs=dictargs, dataframe=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mimetypes\n",
    "import time\n",
    "import os \n",
    "import re\n",
    "import pandas as pd \n",
    "\n",
    "file = 'TEM195660.final.vcf'\n",
    "\n",
    "def assertions(path):\n",
    "    assert os.path.isfile(path), path+\" does'nt exist.\"\n",
    "    return \"[#INFO] File: \"+path\n",
    "    \n",
    "def vcfTodataframe(file, rheader=False):\n",
    "    '''\n",
    "    Take in input vcf file, or tsv and return a dataframe\n",
    "    I\"m gonna build my own vcf parser et puis c'est tout\n",
    "    return 3 Dataframe, full, only sample, only info\n",
    "    '''\n",
    "    print(assertions(file))\n",
    "    name, extension = os.path.splitext(file)\n",
    "\n",
    "    header = []\n",
    "    variants_tmp = []\n",
    "    variants = []\n",
    "    \n",
    "    \n",
    "    if extension == '.vcf':\n",
    "        #print('[#INFO] VCF: '+file)\n",
    "        with open(file) as f:\n",
    "            for lines in f:\n",
    "                if lines[0] == '#':\n",
    "                    header.append(lines.strip())\n",
    "                else:\n",
    "                    variants_tmp.append(lines)\n",
    "                        \n",
    "    else:\n",
    "        #Create a tsv to dataframe, 2lines literally\n",
    "        print(\"ok\")\n",
    "        \n",
    "    print(\"[#INFO]\", header[-1])\n",
    "    #print(variants[-1])\n",
    "   \n",
    "    col = header[-1].strip().split('\\t')\n",
    "    for v in variants_tmp:\n",
    "        variants.append(v.strip().split('\\t'))\n",
    "   \n",
    "    #headerCol = [res.replace('#', '') for res in colTemp]\n",
    "    dfVar = pd.DataFrame(columns=col)\n",
    "    #print(variants[0:3])\n",
    "   \n",
    "    #Creating Dataframe from the whole VCF\n",
    "    print(\"[#INFO] Whole VCF to Dataframe\")\n",
    "    for i, var in enumerate(variants):\n",
    "        print_progress_bar(i, len(variants)-1)\n",
    "        rows = pd.Series(var, index=dfVar.columns)\n",
    "        #print(rows[0])\n",
    "        dfVar = dfVar.append(rows, ignore_index=True)\n",
    "   \n",
    "    print('\\n')\n",
    "    #print(dfVar.head())\n",
    "    #print(dfVar.columns)\n",
    "    if rheader:\n",
    "        return dfVar, header\n",
    "    else: \n",
    "        return dfVar \n",
    "\n",
    "#vcfTodataframe(file)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'vcfTodataframe' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-2-29a0754c757c>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     37\u001b[0m     \u001b[1;32mreturn\u001b[0m \u001b[0mdfInfo\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     38\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 39\u001b[1;33m \u001b[0mDF\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mparseInfoField\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mvcfTodataframe\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m: name 'vcfTodataframe' is not defined"
     ]
    }
   ],
   "source": [
    "import mimetypes\n",
    "import time\n",
    "import os \n",
    "import re\n",
    "import pandas as pd\n",
    "    \n",
    "def parseInfoField(dfVar):\n",
    "    \n",
    "    ############\n",
    "    #Parsing INFO field from dfVar dataframe containing all informations from vcf\n",
    "    ############\n",
    "    \n",
    "    #print(dfVar.head())\n",
    "    infoList = []\n",
    "    dicoInfo = {}\n",
    "    headers = []\n",
    "    \n",
    "    \n",
    "    print(\"[#INFO] Parsing INFO field\")\n",
    "    for i,elems in dfVar.iterrows():\n",
    "        print_progress_bar(i, len(dfVar.index)-1)\n",
    "        infoList.append([pd.Series(x.split('=')) for x in elems['INFO'].split(';')])\n",
    "\n",
    "    print(\"\\n\")\n",
    "    [headers.append(elems[0]) for ite in infoList for elems in ite]\n",
    "    dfInfo = pd.DataFrame(columns=np.unique(np.array(headers)))\n",
    "    \n",
    "    print(\"[#INFO] From INFO field to Dataframe\")\n",
    "    for j, elems in enumerate(infoList):\n",
    "        print_progress_bar(j, len(infoList)-1)\n",
    "        elems = dict(elems)\n",
    "        dfInfo = dfInfo.append(elems, ignore_index=True)\n",
    "    print(\"\\n\")\n",
    "    print(dfInfo.head())\n",
    "    #print(dfInfo.loc['1000genomesAFR'])\n",
    "\n",
    "    return dfInfo\n",
    "\n",
    "DF = parseInfoField(vcfTodataframe(file))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import mimetypes\n",
    "import time\n",
    "import os \n",
    "import re\n",
    "import pandas as pd\n",
    "\n",
    "def parseSampleField(dfVar):\n",
    "    #############\n",
    "    ### Parsing Sample Field in VCF  \n",
    "    #############\n",
    "    \n",
    "    dico = []\n",
    "    #dfTest = pd.Series(dfVar.TEM195660.values,index=dfVar.FORMAT).to_dict()\n",
    "   \n",
    "    #Parsing FORMAT field in VCF\n",
    "    print('\\n')\n",
    "    print(\"[#INFO] Parsing FORMAT field\")\n",
    "    for i, row in dfVar.iterrows():\n",
    "        print_progress_bar(i, len(dfVar.index)-1)\n",
    "        dico.append(pd.Series(row['TEM195660'].split(':'), index=row['FORMAT'].split(':')).to_dict())\n",
    "      \n",
    "    dfSample = pd.DataFrame(dico)\n",
    "    print('\\n')\n",
    "    #print(dfSample.head())\n",
    "    #print('\\n')\n",
    "\n",
    "    return dfSample\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# library & dataset\n",
    "import seaborn as sns\n",
    "df = sns.load_dataset('penguins')\n",
    "import matplotlib.pyplot as plt\n",
    "#print(df.head())\n",
    "#print(type(df))\n",
    "#\n",
    "#inf = parseSampleField(vcfTodataframe(file))\n",
    "#whole = vcfTodataframe(file)\n",
    "#\n",
    "#result = pd.concat([whole, inf], axis=1)\n",
    "#result = result.drop(['FORMAT', 'TEM195660'], axis=1)\n",
    "##dfparsed = dfw.loc['#CHROM', 'AD', 'DP', 'PL']\n",
    "#dfparsed = result.loc[:, ['DP', 'VAF', '#CHROM']]\n",
    "##print(dfparsed.tail())\n",
    "#dfparsed['DP'] = dfparsed['DP'].astype(float)\n",
    "#dfparsed['VAF'] = dfparsed['VAF'].astype(float)\n",
    "#\n",
    "#print(\"longueur:\", len(dfparsed.index))\n",
    "##print(type(dfparsed))\n",
    "##sns.pairplot(dfparsed, hue='#CHROM')\n",
    "\n",
    "#my_list = DF.columns.values.tolist()\n",
    "#print(my_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}